File: cs/Parking/Boy/MostAvailableParkingBoyRule.cs
using System.Collections.Generic;
using System.Linq;

namespace Parking
{
    public class MostAvailableParkingBoyRule : IParkingBoyRule
    {
        public ParkingLot FindOneValidParkingLot(IReadOnlyList<ParkingLot> parkingLots) =>
            parkingLots
            .OrderByDescending(parkingLot => parkingLot.AvailableSpaces)
            .FirstOrDefault(parkingLot => parkingLot.HasAvailableSpaces);
    }
}

File: ts/src/parking/domain/parkingLot/rule/mostAvailableParkingBoyRule.ts
/** @format */

import {ParkingBoyRule} from '../finder/ParkingBoyRule'
import {ParkingLot} from '../parking/parkingLot'

export class MostAvailableParkingBoyRule implements ParkingBoyRule {
    public findOneValidParkingLot(parkingLots: ParkingLot[]) {
        return (
            parkingLots
                .sort(({availableSpaces}) => -availableSpaces)
                .find(({hasAvailableSpaces}) => hasAvailableSpaces) || null
        )
    }
}
